File: packages/node/src/sanitizer.ts
// PRIVACY POLICY: NurseAndrea SDKs never transmit raw env var values,
// connection strings, credentials, hostnames, ports, or API tokens.
// Only derived metadata (type, tech, provider, source, variable_name)
// leaves the host. This module enforces the allowlist.

const ALLOWED_TECHS = new Set([
  "postgresql", "mysql", "sqlite", "mongodb", "redis",
  "memcached", "bullmq", "agenda", "rabbitmq",
])

function normalizeTech(raw: string): string {
  const s = raw.toLowerCase().trim()
  if (s === "postgres" || s === "pg" || s === "postgresql") return "postgresql"
  if (s === "ioredis" || s === "node-redis" || s === "redis") return "redis"
  if (s === "mysql2" || s === "mysql") return "mysql"
  if (s === "sqlite3" || s === "better-sqlite3") return "sqlite"
  if (s === "amqp" || s === "amqps") return "rabbitmq"
  if (s === "mongodb" || s === "mongodb+srv") return "mongodb"
  return s
}

export function sanitizeTech(raw: string | null | undefined): string | null {
  if (!raw) return null
  const normalized = normalizeTech(raw)
  return ALLOWED_TECHS.has(normalized) ? normalized : null
}

export function techFromUrl(url: string): string | null {
  try {
    const scheme = new URL(url).protocol.replace(":", "")
    return sanitizeTech(scheme)
  } catch {
    return null
  }
}

File: packages/ruby/lib/nurse_andrea/sanitizer.rb
# PRIVACY: This file enforces the data privacy policy. Only allowlisted
# fields pass through. See DATA_PRIVACY_POLICY.rb for the full policy.

require "uri"

module NurseAndrea
  module Sanitizer
    DISCOVERY_ALLOWLIST = %i[type tech provider source variable_name].freeze

    def self.sanitize_discovery(raw)
      raw.slice(*DISCOVERY_ALLOWLIST)
    end

    def self.extract_tech(url)
      scheme = URI.parse(url).scheme rescue nil
      case scheme
      when "postgres", "postgresql" then "postgresql"
      when "mysql", "mysql2"        then "mysql"
      when "redis", "rediss"        then "redis"
      when "amqp", "amqps"          then "rabbitmq"
      when "mongodb", "mongodb+srv" then "mongodb"
      else "unknown"
      end
    end

    def self.extract_provider(url)
      host = URI.parse(url).host rescue nil
      return "unknown" unless host

      case host
      when /\.railway\.internal$/       then "railway"
      when /\.render\.com$/             then "render"
      when /\.fly\.dev$/                then "fly"
      when /\.neon\.tech$/              then "neon"
      when /\.supabase\.co$/            then "supabase"
      when /\.upstash\.io$/             then "upstash"
      when /\.mongodb\.net$/            then "mongodb_atlas"
      when /\.herokuapp\.com$/          then "heroku"
      when /\.elephantsql\.com$/        then "elephantsql"
      when /\.aws\.clickhouse\.cloud$/  then "clickhouse_cloud"
      else "self_hosted"
      end
    end
  end
end
